### ISSUE TITLE
Investigate using the WASM Component Model and WIT instead of XDR for contract specs

### ISSUE THREAD

**Issue by leighmcculloch** [stellar/rs-soroban-sdk#683]
WASM Component Model:
- https://github.com/WebAssembly/component-model

WIT file format and related tooling:
- https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md
- https://github.com/bytecodealliance/wit-bindgen

WIT demo generating bindings:
- https://bytecodealliance.github.io/wit-bindgen/

The WASM Component Model project includes the specification of exports from WASM modules and defines a binary format, and a text format (WIT) for describing exports, as well as the wit-bindgen tooling for generating code.

It's not clear to me yet how compatible all elements of what is being built here are compatible or useful to what we're building, however at the very least the ABI binary format and the WIT text format could be a compelling alternative to the XDR based contract specs that we have today. If compelling we could write the binary format to another custom section like we do the XDR form and use it for all the same uses we have today.

The most compelling reasons to investigate this are:
- General compatibility with a growing set of tools in the WASM and Rust ecosystems.

We should explore:
- How stable the binary and text formats are.
- What utilities are available for interacting with them.

cc @graydon I think you looked at the component model early on. I'm curious any commentary you have.
cc @tomerweller Who brought WIT to my attention.

**Comment by tomerweller** [comment]
Inspired by [REAN](https://raen.dev/) which utilizes WIT for Near smart contracts

**Comment by leighmcculloch** [comment]
Another link, there's a `wit-parser` crate inside `wit-bindgen` that provides a parser for the WIT file format.
- https://github.com/bytecodealliance/wit-bindgen/tree/main/crates/wit-parser

**Comment by graydon** [comment]
Yeah I looked at this and "nowhere near done even in spec form, and miles from implemented anywhere" was the understanding I had at the time. It looks like they might be mid way through transition to doing an MVP implementation in wasmtime (see https://github.com/bytecodealliance/wasmtime/issues/4185) and it is .. not wise to ever underestimate Alex Crichton. But I do not see any action in any other wasm implementation repos so I think we're still quite a ways from any of this being really usable. I will ask around to confirm.

**Comment by leighmcculloch** [comment]
@graydon just shared this doc with me, which has an interesting section about the Component Model:
- https://www.assemblyscript.org/standards-objections.html#component-model-2022-09

Specifically it outlines opinions that:
- it is not sufficiently portable because it only "works if all languages involved are Rust-like. Breaks in all other cases."
- it is not sufficiently language neutral as it "only supports one family of programming languages."

**Comment by graydon** [comment]
Again, I'm not opposed to this _in principle_ but we are nowhere near having the support in tools (or even completed specs) to block on this. We can return to it for a v2 interface someday, but it's not happening between now and soroban going live on mainnet.

**Comment by graydon** [comment]
Our timing is terrible .. this might very well be a viable platform to target exactly by the time we ship final. Sigh. Anyway in the meantime wasmi has started shipping some preliminary WASI support -- https://github.com/paritytech/wasmi/blob/23d8d8c684255f2d63526baa348ab3eb3d249de0/CHANGELOG.md?plain=1#L281-L285 -- which is based on the same [wiggle](https://docs.rs/wiggle/7.0.0/wiggle/index.html#) code generator that wasmtime uses to project witx files (the component model's interface definition files) into rust on both host and guest sides.

**Comment by graydon** [comment]
(There is still quite substantial [churn on the ABI](https://github.com/WebAssembly/component-model/commits/main/design/mvp/CanonicalABI.md) so I am not saying we made the wrong decision here, I still think it's not mature enough that could really build on their stuff currently without constant coordination and blocking, but still..)

**Comment by leighmcculloch** [comment]
If we were to explore using the component model, it'd be more than just replacing the contract spec XDR, which is what I implied in this issue's title. It'd require us redoing the entire RawVal<>ScVal type system. Is that correct?

**Comment by graydon** [comment]
I think to a large extent yes, and _probably_ to our detriment, since they aren't optimizing nearly as aggressively for space nor doing anything like a polymorphic value-that-might-also-be-a-host-reference. There are now host reference types in WASM but they're disjoint from other values, so we'd have to put them in an enum / tagged union, which would then be a multi-word value in the projection, AIUI.
